tests/test_default_qubit.py

```diff
@@ -1094,6 +1094,38 @@ def circuit():
 
         assert np.array_equal(outcomes[0], outcomes[1])
 
+    @pytest.mark.parametrize("theta,phi,varphi", list(zip(THETA, PHI, VARPHI)))
+    def test_mutable_qnode(self, theta, phi, varphi, tol):
+        """Test that a mutable QNode evaluated multiple times mutates well and produces
+        the desired result.
+        """
+        dev = qml.device('default.qubit', wires=3)
+        dev.reset()
+
+        def template(weights):
+            qml.RX(weights[0], wires=[0])
+
+        @qml.qnode(dev)
+        def circuit(weights, n_layers=1):
+            for idx in range(n_layers):
+                template(weights[idx])
+            return qml.expval(qml.PauliZ(0))
+
+        weights = np.array([[phi]])
+        res = circuit(weights, n_layers=1)
+        exp = np.cos(phi)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+        weights = np.array([[phi], [theta]])
+        res = circuit(weights, n_layers=2)
+        exp = np.cos(phi+theta)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
+        weights = np.array([[phi], [theta], [varphi]])
+        res = circuit(weights, n_layers=3)
+        exp = np.cos(phi+theta+varphi)
+        assert np.allclose(res, exp, atol=tol, rtol=0)
+
```

Comment:
The simplest test case I could think of that reveals the bug is

```
import pennylane as qml

dev = qml.device("default.qubit", wires=1)

@qml.qnode(dev)
def node(x, n=1):
    for k in range(n):
        qml.RX(x[k], wires=0)
    return qml.expval(qml.PauliZ(wires=0))

node([0.1], n=1)
node([0.1, 0.2], n=2)  # bug: raises an IndexError
```

Reply:
Sure, thanks for that! I'll simplify the way `weights` is handled such that the test is better readable